exec/vector/src/main/java/org/apache/drill/exec/record/metadata/DictBuilder.java

```diff
@@ -0,0 +1,198 @@
+    tupleBuilder.add(DictVector.FIELD_KEY_NAME, type);
+    return this;
+  }
+
+  public DictBuilder addValue(TypeProtos.MinorType type) {
+    tupleBuilder.add(DictVector.FIELD_VALUE_NAME, type);
+    return this;
+  }
+
+  public DictBuilder addKey(TypeProtos.MinorType type, int width) {
+    tupleBuilder.add(DictVector.FIELD_KEY_NAME, type, width);
+    return this;
+  }
+
+  public DictBuilder addValue(TypeProtos.MinorType type, int width) {
+    tupleBuilder.add(DictVector.FIELD_VALUE_NAME, type, width);
+    return this;
+  }
+
+  public DictBuilder addKey(TypeProtos.MinorType type, int precision, int scale) {
+    return addDecimal(DictVector.FIELD_KEY_NAME, type, TypeProtos.DataMode.REQUIRED, precision, scale);
+  }
+
+  public DictBuilder addValue(TypeProtos.MinorType type, int precision, int scale) {
+    return addDecimal(DictVector.FIELD_VALUE_NAME, type, TypeProtos.DataMode.REQUIRED, precision, scale);
+  }
+
+  public DictBuilder addNullable(String name, TypeProtos.MinorType type) {
+    tupleBuilder.addNullable(name,  type);
+    return this;
+  }
+
+  public DictBuilder addNullable(String name, TypeProtos.MinorType type, int width) {
+    tupleBuilder.addNullable(name, type, width);
+    return this;
+  }
+
+  public DictBuilder addNullable(String name, TypeProtos.MinorType type, int precision, int scale) {
+    return addDecimal(name, type, TypeProtos.DataMode.OPTIONAL, precision, scale);
+  }
+
+  public DictBuilder addArrayValue(TypeProtos.MinorType type) {
+    tupleBuilder.addArray(DictVector.FIELD_VALUE_NAME, type);
+    return this;
+  }
+
+  public DictBuilder addArrayValue(TypeProtos.MinorType type, int dims) {
+    tupleBuilder.addArray(DictVector.FIELD_VALUE_NAME, type, dims);
+    return this;
+  }
+
+  public DictBuilder addArrayValue(TypeProtos.MinorType type, int precision, int scale) {
+    return addDecimal(DictVector.FIELD_VALUE_NAME, type, TypeProtos.DataMode.REPEATED, precision, scale);
+  }
+
+  public DictBuilder addDecimal(String name, TypeProtos.MinorType type,
+                               TypeProtos.DataMode mode, int precision, int scale) {
+    tupleBuilder.addDecimal(name, type, mode, precision, scale);
+    return this;
+  }
+
+  /**
+   * Add a map column as dict's value. The returned schema builder is for the nested
+   * map. Building that map, using {@link DictBuilder#resumeSchema()},
+   * will return the original schema builder.
+   *
+   * @return a builder for the map
+   */
+  public MapBuilder addMapValue() {
+    return tupleBuilder.addMap(this, DictVector.FIELD_VALUE_NAME);
+  }
+
+  public MapBuilder addMapArrayValue() {
+    return tupleBuilder.addMapArray(this, DictVector.FIELD_VALUE_NAME);
+  }
+
+  public DictBuilder addDictValue() {
+    return tupleBuilder.addDict(this, DictVector.FIELD_VALUE_NAME);
+  }
+
+  public UnionBuilder addUnionValue() {
+    return tupleBuilder.addUnion(this, DictVector.FIELD_VALUE_NAME);
+  }
+
+  public UnionBuilder addListValue() {
```

Comment:
Again, if we can only have one value type, maybe use `listValue()`, etc. instead of `addListValue()`.